core/src/main/java/io/grpc/DnsNameResolver.java

```diff
@@ -94,7 +117,7 @@ public void run() {
         try {
           inetAddrs = InetAddress.getAllByName(host);
         } catch (Exception e) {
-          listener.onError(Status.UNAVAILABLE.withCause(e));
+          savedListener.onError(Status.UNAVAILABLE.withCause(e));
           return;
```

Comment:
resolving = false isn't done in this code path. You may consider a try-finally (although it is unfortunate it would increase the indentation).